text/0000-erfc-post-build-contexts.md

```diff
@@ -0,0 +1,396 @@
+attributes, in which case it would get all items with any of the
+listed attributes. These items be modules, functions, structs,
+statics, or whatever else the post-build context wants to support. Note
+that the post-build context function can only see all the annotated
+items, not modify them; modification would have to happen with regular
+procedural macros. The returned `TokenStream` must declare the `main()`
+that is to become the entry-point for the binary produced when this
+post-build context is used.
+
+Because this procedural macro is only loaded when it is used as the
+post-build context, the `#[mytest]` annotation should probably be kept
+behind `#[cfg(mytest)]` (which is automatically set when the `mytest`
+context is used) so that you don't get unknown attribute warnings
+whilst loading, and to avoid conflicts with other post-build contexts
+that may use the same attributes. (We could change this by asking
+attributes to be registered in Cargo.toml, but we don't find this
+necessary)
+
+## Cargo integration
+
+Alternative post-build contexts need to integrate with cargo.
+In particular, when crate `a` uses a crate `b` which provides an
+post-build context, `a` needs to be able to specify when `b`'s post-build
+context should be used. Furthermore, cargo needs to understand that when
+`b`'s post-build context is used, `b`'s dependencies must also be linked.
+Note that `b` could potentially provide multiple post-build contexts ---
+these are named according to the name of their `#[post_build_context]`
+function.
+
+Crates which define a post-build context must have an `post-build-context = true`
+key in their `Cargo.toml`. It can also specify
+
+```rust
+single-target = true    # false by default
+```
+
+`single-target` indicates that only a single target can be run with this
+context at once (some tools, like cargo-fuzz, run forever, and so it
+does not make sense to specify multiple targets).
+
+Crates that wish to *use* a custom post-build context, do so by defining
+a new post-build context under a new `build-context` section in their
+`Cargo.toml`:
+
+```toml
+[post-build.context.fuzz]
+provider = { rust-fuzz = "1.0" }
+folders = ["fuzz/"]
+```
+
+This defines a post-build context named `fuzz`, which uses the
+implementation provided by the `rust-fuzz` crate. When run, it will be
+applied to all files in the `fuzz` directory. By default, the following
+contexts are defined:
+
+```toml
+[post-build.context.test]
+provider = { test = "1.0", context = "test" }
+folders = ["tests/"]
+
+[post-build.context.bench]
+provider = { test = "1.0", context = "bench" }
+folders = ["benchmarks/"]
+```
+
+There's also an `example` context defined that just runs the `main()` of
+any files given.
+
+These can be overridden by a crate's `Cargo.toml`. The `context`
+property is used to disambiguate when a single crate has multiple
+functions tagged `#[post_build_context]` (if we were using the example
```

Comment:
wfm

Reply:
Note that libtest defines two. But if we're moving bench out anyway, that's fine.

Reply:
(decided to instead change it so that once libtest starts using this, we separate out bench and example into their separate PBC crates, if we keep bench around at all)

Reply:
Yeah, I think bench should not be in libtest